proposals/async-method-builders.md

```diff
@@ -46,23 +46,70 @@ namespace System.Runtime.CompilerServices
     /// build the attributed method.
     /// </summary>
-    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Struct | AttributeTargets.Interface | AttributeTargets.Method | AttributeTargets.Module, Inherited = false, AllowMultiple = false)]
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Struct | AttributeTargets.Interface | AttributeTargets.Method | AttributeTargets.Module, Inherited = false, AllowMultiple = true)]
     public sealed class AsyncMethodBuilderOverrideAttribute : Attribute
     {
         /// <summary>Initializes the <see cref="AsyncMethodBuilderOverrideAttribute"/>.</summary>
         /// <param name="builderType">The <see cref="Type"/> of the associated builder.</param>
         public AsyncMethodBuilderOverrideAttribute(Type builderType) => BuilderType = builderType;
 
-        // for scoped application (use property for targetReturnType? have compiler check that it's provided)
-        public AsyncMethodBuilderOverrideAttribute(Type builderType, Type targetReturnType) => ...
-
         /// <summary>Gets the <see cref="Type"/> of the associated builder.</summary>
         public Type BuilderType { get; }
     }
 }
 ```
 
-In the C# compiler, prefer the attribute on the method when determining what builder to use over the one defined on the type.  For example, today if a method is defined as:
+The attribute can be applied on methods, types and modules.
+
+Example of usage on a method:  
+```C#
+[AsyncMethodBuilderOverride(typeof(PoolingAsyncValueTaskMethodBuilder<int>))] // new, referring to some custom builder type
+static async ValueTask<int> ExampleAsync() { ... }
+```
+
+A developer that wants to using a specific custom builder for all of their methods can do so by putting the relevant attribute on each method.  
+Example of usage on module:  
+```C#
+[module: AsyncMethodBuilderOverride(typeof(PoolingAsyncValueTaskMethodBuilder))]
+[module: AsyncMethodBuilderOverride(typeof(PoolingAsyncValueTaskMethodBuilder<>))]
+
+class MyClass
+{
+    public async ValueTask Method1Async() { ... } // would use PoolingAsyncValueTaskMethodBuilder
+    public async ValueTask<int> Method2Async() { ... } // would use PoolingAsyncValueTaskMethodBuilder<int>
+    public async ValueTask<string> Method3Async() { ... } // would use PoolingAsyncValueTaskMethodBuilder<string>
+}
+```
+
+### Determining the builder type for an async method
+
+When compiling an async method, the builder type is determined by:
+1. looking in the containing scopes for an override attribute that specifies a builder type compatible with the method's return type.
+2. otherwise, falling back to the builder type determined by previous approach. (see [spec for task-like types](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-7.0/task-types.md)) 
+
+The containing scopes are considered starting from the method, then its containing types and finally the module. Each scope is considered
```

Comment:
Couple of items that I think we should add clarity on: 

- How does this work for local functions? Does a local function consider the attributes on the function that contains it or is it containing scope the type?
- By containing types here we mean we consider the inner most nested type and walk outwards to the containing type correct?

Reply:
- Good catch for local functions and lambdas. I'm proposing that the attribute also apply to those and those scopes contribute to the lookup chain.

- Yes, we'll look at the containing types in turn, walking outwards.